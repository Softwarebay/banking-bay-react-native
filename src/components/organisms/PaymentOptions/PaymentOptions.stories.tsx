import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react-native';
import { CenterStory } from 'utils/storybook/CenterStory';

import { PaymentOptions } from './PaymentOptions';

storiesOf('Organisms/PaymentOptions', module)
	.addDecorator((getStory) => <CenterStory>{getStory()}</CenterStory>)
	.add('Default', () => (
		<PaymentOptions
			onPressMakePayment={action('onPressMakePayment')}
			onPressMobilePayment={action('onPressMobilePayment')}
			onPressMoneyTransfer={action('onPressMoneyTransfer')}
			onPressTopUpPayment={action('onPressTopUpPayment')}
		/>
	));
