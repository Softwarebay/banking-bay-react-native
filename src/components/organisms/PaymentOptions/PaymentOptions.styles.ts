import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
	btn: {
		borderRadius: 100,
		height: 60,
		marginBottom: 6,
		width: 60,
	},
	option: {
		height: '100%',
	},
	optionTitle: {
		fontSize: 10,
		fontWeight: '600',
		lineHeight: 12,
		maxWidth: 60,
		textAlign: 'center',
	},
	options: {
		justifyContent: 'space-around',
	},
});
