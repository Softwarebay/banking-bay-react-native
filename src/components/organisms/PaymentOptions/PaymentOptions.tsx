import { View } from 'react-native';

import { Flex, IconButton, IconProps, Text } from 'components/atoms';

import { colors } from 'theme';
import { styles } from './PaymentOptions.styles';

export interface PaymentOption {
	id: string;
	title: string;
	btnColor: string;
	iconColor: string;
	iconName: IconProps['name'];
	onPress: () => void;
}

export interface Props {
	onPressTopUpPayment: () => void;
	onPressMobilePayment: () => void;
	onPressMoneyTransfer: () => void;
	onPressMakePayment: () => void;
}

export const PaymentOptions = ({
	onPressTopUpPayment,
	onPressMobilePayment,
	onPressMoneyTransfer,
	onPressMakePayment,
}: Props) => {
	const paymentOptions: PaymentOption[] = [
		{
			id: 'top-up-payment',
			title: 'Top-Up Payment',
			btnColor: '#3EB290',
			iconColor: colors.white,
			iconName: 'credit-card',
			onPress: onPressTopUpPayment,
		},
		{
			id: 'mobile-payment',
			title: 'Mobile Payment',
			btnColor: '#FF8A71',
			iconColor: colors.white,
			iconName: 'smartphone',
			onPress: onPressMobilePayment,
		},
		{
			id: 'money-transfer',
			title: 'Money Transfer',
			btnColor: '#55ACEE',
			iconColor: colors.white,
			iconName: 'repeat',
			onPress: onPressMoneyTransfer,
		},
		{
			id: 'make-payment',
			title: 'Make a Payment',
			btnColor: '#EECC55',
			iconColor: colors.white,
			iconName: 'dollar-sign',
			onPress: onPressMakePayment,
		},
	];

	return (
		<Flex direction="row" style={styles.options}>
			{paymentOptions.map(
				({ id, btnColor, iconName, iconColor, title, onPress }) => (
					<View key={id} style={styles.option}>
						<IconButton
							btnStyle={[styles.btn, { backgroundColor: btnColor }]}
							iconColor={iconColor}
							iconName={iconName}
							iconSize={25}
							testID={`${id}-icon-btn-test-id`}
							onPress={onPress}
						/>
						<Text
							style={styles.optionTitle}
							testID={`${id}-title-test-id`}
							text={title}
						/>
					</View>
				)
			)}
		</Flex>
	);
};
