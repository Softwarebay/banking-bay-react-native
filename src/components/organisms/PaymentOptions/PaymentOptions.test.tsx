import { fireEvent, render, screen } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { PaymentOptions } from './PaymentOptions';

import type { Props } from './PaymentOptions';

describe('PaymentOptions', () => {
	const mockOnPressMakePayment = jest.fn<Props['onPressMakePayment'], []>();
	const mockOnPressMobilePayment = jest.fn<Props['onPressMobilePayment'], []>();
	const mockOnPressMoneyTransfer = jest.fn<Props['onPressMoneyTransfer'], []>();
	const mockOnPressTopUpPayment = jest.fn<Props['onPressTopUpPayment'], []>();

	const props: Props = {
		onPressMakePayment: mockOnPressMakePayment,
		onPressMobilePayment: mockOnPressMobilePayment,
		onPressMoneyTransfer: mockOnPressMoneyTransfer,
		onPressTopUpPayment: mockOnPressTopUpPayment,
	};

	it('should match snapshot', () => {
		const tree = testRenderer.create(<PaymentOptions {...props} />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should render 4 icon buttons with bottom text', () => {
		render(<PaymentOptions {...props} />);

		const iconButtons = screen.getAllByTestId(
			/(top-up-payment|mobile-payment|money-transfer|make-payment)-icon-btn-test-id/gi
		);

		expect(iconButtons).toHaveLength(4);

		const makePaymentIcon = screen.getByTestId('dollar-sign-icon-test-id');
		const moneyTransferIcon = screen.getByTestId('repeat-icon-test-id');
		const topUpPaymentIcon = screen.getByTestId('credit-card-icon-test-id');
		const mobilePaymentIcon = screen.getByTestId('smartphone-icon-test-id');

		expect(makePaymentIcon).not.toBeNull();
		expect(moneyTransferIcon).not.toBeNull();
		expect(topUpPaymentIcon).not.toBeNull();
		expect(mobilePaymentIcon).not.toBeNull();

		const bottomTextElements = screen.getAllByTestId(
			/(top-up-payment|mobile-payment|money-transfer|make-payment)-title-test-id/gi
		);

		const bottomTexts = bottomTextElements.map((el) => el.children[0]);

		expect(bottomTexts).toContain('Make a Payment');
		expect(bottomTexts).toContain('Money Transfer');
		expect(bottomTexts).toContain('Top-Up Payment');
		expect(bottomTexts).toContain('Mobile Payment');
	});

	it("should handle press events on 'Make a Payment' icon button", () => {
		render(<PaymentOptions {...props} />);

		const makePaymentIconBtn = screen.getByTestId(
			'make-payment-icon-btn-test-id'
		);

		fireEvent.press(makePaymentIconBtn);

		expect(mockOnPressMakePayment).toHaveBeenCalledTimes(1);
		expect(mockOnPressMobilePayment).not.toHaveBeenCalled();
		expect(mockOnPressMoneyTransfer).not.toHaveBeenCalled();
		expect(mockOnPressTopUpPayment).not.toHaveBeenCalled();
	});

	it("should handle press events on 'Mobile Payment' icon button", () => {
		render(<PaymentOptions {...props} />);

		const mobilePaymentIconBtn = screen.getByTestId(
			'mobile-payment-icon-btn-test-id'
		);

		fireEvent.press(mobilePaymentIconBtn);

		expect(mockOnPressMobilePayment).toHaveBeenCalledTimes(1);
		expect(mockOnPressMakePayment).not.toHaveBeenCalled();
		expect(mockOnPressMoneyTransfer).not.toHaveBeenCalled();
		expect(mockOnPressTopUpPayment).not.toHaveBeenCalled();
	});

	it("should handle press events on 'Money Transfer' icon button", () => {
		render(<PaymentOptions {...props} />);

		const moneyTransferIconBtn = screen.getByTestId(
			'money-transfer-icon-btn-test-id'
		);

		fireEvent.press(moneyTransferIconBtn);

		expect(mockOnPressMoneyTransfer).toHaveBeenCalledTimes(1);
		expect(mockOnPressMakePayment).not.toHaveBeenCalled();
		expect(mockOnPressMobilePayment).not.toHaveBeenCalled();
		expect(mockOnPressTopUpPayment).not.toHaveBeenCalled();
	});

	it("should handle press events on 'Top-Up Payment' icon button", () => {
		render(<PaymentOptions {...props} />);

		const topUpPaymentIconBtn = screen.getByTestId(
			'top-up-payment-icon-btn-test-id'
		);

		fireEvent.press(topUpPaymentIconBtn);

		expect(mockOnPressTopUpPayment).toHaveBeenCalledTimes(1);
		expect(mockOnPressMakePayment).not.toHaveBeenCalled();
		expect(mockOnPressMobilePayment).not.toHaveBeenCalled();
		expect(mockOnPressMoneyTransfer).not.toHaveBeenCalled();
	});
});
