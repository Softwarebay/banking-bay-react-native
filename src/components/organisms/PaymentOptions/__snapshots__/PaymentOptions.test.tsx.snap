// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentOptions should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
      },
      Object {
        "flexDirection": "row",
      },
      Object {
        "justifyContent": "space-around",
      },
    ]
  }
>
  <View
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={20}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
          Array [
            Object {
              "borderRadius": 100,
              "height": 70,
              "marginBottom": 10,
              "width": 70,
            },
            Object {
              "backgroundColor": "#3EB290",
            },
          ],
          false,
        ]
      }
      testID="top-up-payment-icon-btn-test-id"
    >
      <SvgMock
        color="#FFFFFF"
        height={25}
        testID="credit-card-icon-test-id"
        width={25}
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#4C4C60",
            "fontFamily": "MulishRegular",
            "fontSize": 17,
            "lineHeight": 27.2,
          },
          Object {
            "fontFamily": "MulishSemiBold",
            "fontSize": 12,
            "lineHeight": 14.4,
            "maxWidth": 70,
            "textAlign": "center",
          },
        ]
      }
      testID="top-up-payment-title-test-id"
    >
      Top-Up Payment
    </Text>
  </View>
  <View
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={20}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
          Array [
            Object {
              "borderRadius": 100,
              "height": 70,
              "marginBottom": 10,
              "width": 70,
            },
            Object {
              "backgroundColor": "#FF8A71",
            },
          ],
          false,
        ]
      }
      testID="mobile-payment-icon-btn-test-id"
    >
      <SvgMock
        color="#FFFFFF"
        height={25}
        testID="smartphone-icon-test-id"
        width={25}
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#4C4C60",
            "fontFamily": "MulishRegular",
            "fontSize": 17,
            "lineHeight": 27.2,
          },
          Object {
            "fontFamily": "MulishSemiBold",
            "fontSize": 12,
            "lineHeight": 14.4,
            "maxWidth": 70,
            "textAlign": "center",
          },
        ]
      }
      testID="mobile-payment-title-test-id"
    >
      Mobile Payment
    </Text>
  </View>
  <View
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={20}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
          Array [
            Object {
              "borderRadius": 100,
              "height": 70,
              "marginBottom": 10,
              "width": 70,
            },
            Object {
              "backgroundColor": "#55ACEE",
            },
          ],
          false,
        ]
      }
      testID="money-transfer-icon-btn-test-id"
    >
      <SvgMock
        color="#FFFFFF"
        height={25}
        testID="repeat-icon-test-id"
        width={25}
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#4C4C60",
            "fontFamily": "MulishRegular",
            "fontSize": 17,
            "lineHeight": 27.2,
          },
          Object {
            "fontFamily": "MulishSemiBold",
            "fontSize": 12,
            "lineHeight": 14.4,
            "maxWidth": 70,
            "textAlign": "center",
          },
        ]
      }
      testID="money-transfer-title-test-id"
    >
      Money Transfer
    </Text>
  </View>
  <View
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={20}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
          Array [
            Object {
              "borderRadius": 100,
              "height": 70,
              "marginBottom": 10,
              "width": 70,
            },
            Object {
              "backgroundColor": "#EECC55",
            },
          ],
          false,
        ]
      }
      testID="make-payment-icon-btn-test-id"
    >
      <SvgMock
        color="#FFFFFF"
        height={25}
        testID="dollar-sign-icon-test-id"
        width={25}
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#4C4C60",
            "fontFamily": "MulishRegular",
            "fontSize": 17,
            "lineHeight": 27.2,
          },
          Object {
            "fontFamily": "MulishSemiBold",
            "fontSize": 12,
            "lineHeight": 14.4,
            "maxWidth": 70,
            "textAlign": "center",
          },
        ]
      }
      testID="make-payment-title-test-id"
    >
      Make a Payment
    </Text>
  </View>
</View>
`;
